src/Cmfd.cpp

```diff
@@ -394,54 +402,272 @@ void Cmfd::setNumDomains(int num_x, int num_y, int num_z) {
+               added_cells_y.c_str());
+  if (modified_cells_y.compare("") != 0)
+    log_printf(WARNING_ONCE, "CMFD mesh cell widths adjusted to fit "
+               "domain decomposition boundaries in the Y direction "
+               "at [%s] cm.", modified_cells_y.c_str());
+
+  /* Z-direction */
+  std::string added_cells_z = "";
+  std::string modified_cells_z = "";
+  cells_added = 0;
+  for (iter = divisions_missing_z.begin(); iter != divisions_missing_z.end();
+       ++iter) {
+
+    j_prev = iter->first;
+    double cmfd_cell_width = _accumulate_z[j_prev+1] - _accumulate_z[j_prev];
+    double coord = iter->second;
+
+    /* If the new CMFD cells are about to be very small,
+       move the CMFD cell boundaries instead, to the closest one */
+    double delta_below = coord - _accumulate_z[j_prev];
+    double delta_above = _accumulate_z[j_prev+1] - coord;
+
+    if (delta_below < 0.1 * cmfd_cell_width ||
+        delta_above < 0.1 * cmfd_cell_width) {
+
+      modified_cells_z += (std::to_string(coord) + " ");
+      if (delta_below < delta_above) {
+        _cell_widths_z[j_prev + cells_added - 1] += delta_below;
+        _cell_widths_z[j_prev + cells_added] -= delta_below;
+      }
+      else {
+        _cell_widths_z[j_prev + cells_added] -= delta_above;
+        _cell_widths_z[j_prev + cells_added + 1] += delta_above;
+        //FIXME Capture j_prev+1 overflow
+      }
+    }
+
+    /* Else, add a new subdivision to the CMFD mesh */
+    else {
+      _cell_widths_z[j_prev + cells_added] = delta_below;
+      _cell_widths_z.insert(_cell_widths_z.begin() + j_prev + cells_added + 1,
+                            delta_above);
+      added_cells_z += (std::to_string(coord) + " ");
+      cells_added++;
+    }
+  }
+  if (added_cells_z.compare("") != 0)
+    log_printf(WARNING_ONCE, "New CMFD cells created for domain decomposition "
+               "boundaries in the Z direction at [%s] cm.",
+               added_cells_z.c_str());
+  if (modified_cells_z.compare("") != 0)
+    log_printf(WARNING_ONCE, "CMFD mesh cell widths adjusted to fit "
+               "domain decomposition boundaries in the Z direction "
+               "at [%s] cm.", modified_cells_z.c_str());
+
+  /* Re-initialize lattice since widths have been modified */
+  if (divisions_missing_x.size() > 0 || divisions_missing_y.size() > 0 ||
+      divisions_missing_z.size() > 0) {
+    _non_uniform = true;
+    Point offset;
+    offset.setXYZ(_lattice->getOffset()->getXYZ());
+    initializeLattice(&offset);
+
+    /* Re-run routine to obtain _accumulate_lmxyz arrays */
+    if (!_widths_adjusted_for_domains) {
+      _widths_adjusted_for_domains = true;
+      setNumDomains(num_x, num_y, num_z);
+    }
+    else
+      log_printf(ERROR, "Mesh adjustment to domain decomposition boundaries "
+                 "failed. Manually add the domain boundaries in a non-"
+                 "uniform CMFD mesh.");
+
+  }
```

Comment:
It is awesome to have this function CUZ it is annoying to adjusting the boundaries manually. Do you have any unit test for it?

Reply:
I should but setDomainDecomposition() is not included in the regular build (without -MPIx)
I'll include one when we start testing the domain decomposed solver. We have a UROP  working on improving the test suite this semester